internal/controller/ngrok/cloudendpoint_controller.go

```diff
@@ -0,0 +1,342 @@
+
+// #region Reconcile CRUD
+
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=cloudendpoints,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=cloudendpoints/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=cloudendpoints/finalizers,verbs=update
+
+func (r *CloudEndpointReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	clep := new(ngrokv1alpha1.CloudEndpoint)
+
+	// Call the base controller's Reconcile method
+	result, err := r.controller.Reconcile(ctx, req, clep)
+
+	// Handle the custom error for domain creation delay
+	if errors.Is(err, ErrDomainCreating) {
+		// Requeue after 10 seconds to allow time for domain registration
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	return result, err
+}
+
+// Create will make sure a domain is created before creating the Cloud Endpoint
+// It also looks up the Traffic Policy and creates the Cloud Endpoint using this Traffic Policy JSON
+func (r *CloudEndpointReconciler) create(ctx context.Context, clep *ngrokv1alpha1.CloudEndpoint) error {
+	err := r.ensureDomainExists(ctx, clep)
+	if err != nil {
+		return err
+	}
+
+	policy, err := r.findTrafficPolicy(ctx, clep.Spec.TrafficPolicyName, clep.Namespace)
+	if err != nil {
+		return err
+	}
+
+	createParams := &ngrok.EndpointCreate{
+		Type:          "cloud",
+		URL:           clep.Spec.URL,
+		Description:   &clep.Spec.Description,
+		Metadata:      &clep.Spec.Metadata,
+		TrafficPolicy: policy,
+		Bindings:      clep.Spec.Bindings,
+	}
+
+	ngrokClep, err := r.NgrokClientset.Endpoints().Create(ctx, createParams)
+	if err != nil {
+		return err
+	}
+
+	return r.updateStatus(ctx, clep, ngrokClep)
+}
+
+// Update is called when we have a status ID and want to update the resource in the ngrok API
+// If it fails to find the resource by ID, create a new one instead
+func (r *CloudEndpointReconciler) update(ctx context.Context, clep *ngrokv1alpha1.CloudEndpoint) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("id", clep.Status.ID)
+
+	err := r.ensureDomainExists(ctx, clep)
+	if err != nil {
+		return err
+	}
+
+	policy, err := r.findTrafficPolicy(ctx, clep.Spec.TrafficPolicyName, clep.Namespace)
+	if err != nil {
+		return err
+	}
+
+	updateParams := &ngrok.EndpointUpdate{
+		ID:            clep.Status.ID,
+		Url:           &clep.Spec.URL,
+		Description:   &clep.Spec.Description,
+		Metadata:      &clep.Spec.Metadata,
+		TrafficPolicy: &policy,
+		Bindings:      clep.Spec.Bindings,
+	}
+
+	ngrokClep, err := r.NgrokClientset.Endpoints().Update(ctx, updateParams)
+	if ngrok.IsNotFound(err) {
+		// Couldn't find endpoint by ID to update, so blank it out and create a new one
+		log.Info("Failed to update endpoint %s by ID because it was not found. Creating a new one", clep.Status.ID)
+		clep.Status.ID = ""
+		_ = r.Client.Status().Update(ctx, clep)
+		return r.create(ctx, clep)
```

Comment:
that works, probably makes sure we get consistent logs and events by going through the re-reconciler

Reply:
actually, if i just return the error here, I don't think we'll actually requeue it since in the base controller it will see it as an ngrok error that's not a 500 or too many requests. I'm not sure we always want to default to a retry every ngrok 404. I could update the base controllers reconcile to check for errors on update and if its an ngrok 404 to do this requeuing if you want me to try to solve it in a dry way.